src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,378 @@
+}
+
+static int scheduler_dp_task_cancel(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_CANCEL;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	return 0;
+}
+
+static int scheduler_dp_task_complete(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+	struct task_dp_pdata *pdata = task->priv_data;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_COMPLETED;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	/* let the thread run once again to call task_complete in a thread context */
+	k_sem_give(&pdata->sem);
+
+	return 0;
+}
+
+static int scheduler_dp_task_free(void *data, struct task *task)
+{
+	k_spinlock_key_t lock_key;
+	struct task_dp_pdata *pdata = task->priv_data;
+
+	/* abort the execution of the thread */
+	k_thread_abort(pdata->thread_id);
+
+	lock_key = scheduler_dp_lock();
+	list_item_del(&task->list);
+	task->priv_data = NULL;
+	task->state = SOF_TASK_STATE_FREE;
+	scheduler_dp_unlock(lock_key);
+
+	/* free all allocated resources */
+	rfree(pdata->p_stack);
+	rfree(pdata->thread_id);
+	rfree(pdata);
+
+	return 0;
+}
+
+/* Thread function called in component context, on target core */
+static void dp_thread_fn(void *p1, void *p2, void *p3)
+{
+	struct task *task = p1;
+	(void) p2;
+	(void) p3;
+	struct task_dp_pdata *task_pdata = task->priv_data;
+	k_spinlock_key_t lock_key;
+	enum task_state state;
+
+	while (1) {
+		/*
+		 * the thread is started immediately after creation, it will stop on semaphore
+		 * Semaphore will be released once the task is ready to process
+		 */
+		k_sem_take(&task_pdata->sem, K_FOREVER);
+
+		if (task->state == SOF_TASK_STATE_RUNNING)
+			state = task_run(task);
+		else
+			state = task->state;	/* to avoid undefined variable warning */
+
+		lock_key = scheduler_dp_lock();
+		/*
+		 * check if task is still running, may have been canceled by external call
+		 * if not, set the state returned by run procedure
+		 */
+		if (task->state == SOF_TASK_STATE_RUNNING) {
+			task->state = state;
+			switch (state) {
+			case SOF_TASK_STATE_RESCHEDULE:
+				/* mark to reschedule, schedule time is already calculated */
+				task->state = SOF_TASK_STATE_QUEUED;
+				break;
+
+			case SOF_TASK_STATE_CANCEL:
+			case SOF_TASK_STATE_COMPLETED:
+				/* remove from scheduling */
+				list_item_del(&task->list);
+				break;
+
+			default:
+				/* illegal state */
+				assert(false);
+			}
+
+		}
+
+		/* call task_complete  */
+		if (task->state == SOF_TASK_STATE_COMPLETED) {
+			/*
+			 * clear semaphore count to avoid executing task_complete twice
+			 * in case the task_complete have been called during task execution
+			 */
+			k_sem_reset(&task_pdata->sem);
```

Comment:
This looks racy.  The k_sem_give() call does not seem to be done with the spinlock held, so if you reach this at the same time the "tick" code is trying to wake you up for the next event, you'll lose the update and sleep forever.

Basically: the counting behavior of semaphores is your friend and prevents this kind of race.  If you need to reset it in a context that's not some kind of error handling, it's probably because there's a bug with the state management (e.g. why was "task_complete() called during task execution", isn't that a bug? If not, shouldn't task_complete() be responsible for squashing the duplicates?)

Reply:
Not quite - I DO want to sleep forever when task is COMPLETED. Time tick won't ever wake it up. The only way to wake the task is to call task_schedule again

the situation I want to **prevent is double call of task_complete** - 
 - task is running (semaphore locked)
 - call to task complete arrive (async) => unlock the semaphore to ensure task_complete is called in task context
 - task_complete is called
 - the loop goes to start - semaphore is unlocked
 **- task_complete is called again :(**
 - task sleeps forever (as designed) on semaphore

ANYWAY - task completed is to be removed. So this prevention won't be needed

Reply:
task complete removed, semaphore reset not needed anymore